smart_simulation/scale.py

```diff
@@ -0,0 +1,235 @@
+
+daiquiri.setup(
+    level=logging.INFO,
+    outputs=(
+        daiquiri.output.Stream(
+            formatter=daiquiri.formatter.ColorFormatter(
+                fmt="%(asctime)s [%(levelname)s] %(name)s.%(" "funcName)s: %(message)s"
+            )
+        ),
+    ),
+)
+
+
+def upsample(dataset, consumption_window_template: str) -> pd.DataFrame:
+    """
+    Upsample a dataset of daily servings to a specified frequency and distribute servings based on configured
+    probabilities
+
+    Args:
+        dataset: Dataset of time stamps and servings (consumed) by consumer
+        consumption_window_template: Upsampled time windows and probabilities
+
+    Returns:
+        Upsampled data as a Pandas DataFrame
+
+    """
+
+    consumption_window = auxiliary.consumption_windows[consumption_window_template]
+    time_stamps = list(consumption_window.keys())
+    weights = list(consumption_window.values())
+    frequency = pd.infer_freq(time_stamps)
+
+    upsampled_time_stamps = pd.date_range(
+        start=dataset.index[0],
+        end=(dataset.index[-1] + pd.Timedelta("1 days") - pd.Timedelta(frequency)),
+        freq=frequency,
+    )
+
+    upsampled_data = pd.DataFrame(index=upsampled_time_stamps).assign(servings=0)
+
+    for day in dataset.index:
+        for serving in range(dataset.loc[day]["servings"]):
+            period = random.choices(population=time_stamps, weights=weights, k=1)[0]
+            time_stamp = day + pd.to_timedelta(period)
+            upsampled_data.at[time_stamp, "servings"] += 1
+
+    return upsampled_data
+
+
+def create_weight_data(dataset: pd.DataFrame, arrival_template: str) -> pd.DataFrame:
+    """
+    Create weight data based on consumption and shipment arrivals
+    Args:
+        dataset: Dataset of time stamps and servings (consumed) by consumer
+        arrival_template: Arrival windows and probabilities used for random selection of new product arrival on scale
+
+    Returns:
+        Pandas DataFrame of weight measurements
+
+    """
+    arrival_options = auxiliary.arrival_options[arrival_template]
+    arrival_options_weights = tuple(
+        auxiliary.arrival_options_weights[arrival_template].values()
+    )
+    stock_weight = auxiliary.stock_weight
+    quantity_to_weight = auxiliary.quantity_to_weight
+    data = dataset.copy()
+    data = data.assign(weight=0.0)
+    weight_col = data.columns.get_loc("weight")
+    servings_col = data.columns.get_loc("servings")
+
+    data.iat[0, weight_col] = stock_weight
+
+    time_step = 1
+    while time_step < len(data.index):
+        previous_step_weight = data.iat[time_step - 1, weight_col]
+        consumption_servings = data.iat[time_step, servings_col]
+
+        if previous_step_weight == 0.0:
+            resupply_arrival = random.choices(
+                population=[*arrival_options], weights=arrival_options_weights, k=1
+            )[0]
+            arrival_timing = arrival_options[resupply_arrival][0](
```

Comment:
I'd probably rewrite this to use more lines of code for more clarify. It's a lot to parse mentally.